lib/decompress/zstd_decompress_block.c

```diff
@@ -974,11 +974,20 @@ ZSTD_decompressSequences_body( ZSTD_DCtx* dctx,
                 BIT_DStream_endOfBuffer < BIT_DStream_completed &&
                 BIT_DStream_completed < BIT_DStream_overflow);
 
+#ifdef __aarch64__
+        BIT_reloadUnfinishedDStream(&(seqState.DStream));
+#else
+        BIT_reloadDStream(&(seqState.DStream));
+#endif
         for ( ; ; ) {
             seq_t const sequence = ZSTD_decodeSequence(&seqState, isLongOffset, ZSTD_p_noPrefetch);
             size_t const oneSeqSize = ZSTD_execSequence(op, oend, sequence, &litPtr, litEnd, prefixStart, vBase, dictEnd);
             DEBUGLOG(6, "regenerated sequence size : %u", (U32)oneSeqSize);
+#ifdef __aarch64__
+            BIT_reloadUnfinishedDStream(&(seqState.DStream));
```

Comment:
I think this is incorrect. This could cause zstd to read out of bounds on corrupted inputs, since we’re removing the bounds check. I’m only mostly certain, but I remember looking into a similar change, and realizing it is wrong.

Reply:
@terrelln Thanks. Is the test result related to the test platform or test data? I executed “make test” on aarch64 and x86, and no error was reported. Are there other methods to check whether my modification is correct?

Reply:
This test result https://travis-ci.org/facebook/zstd/jobs/645820930?utm_medium=notification&utm_source=github_status is an out of bounds access caused by this PR, probably this line.

You can run the fuzzers following this README https://github.com/facebook/zstd/tree/dev/tests/fuzz. The following is about what you need to do to run the `simple_decompress` fuzzer which is the most relevant.

```
cd tests/fuzz
make corpora # download corpus
CC=clang CXX=clang++ ./fuzz.py build --enable-fuzzer --enable-asan --enable-ubsan
./fuzz.py libfuzzer simple_decompress
```

You should also run zstreamtest and fuzzer. In this example they run for 3 minutes after running the "unit tests". These will be a faster turn around than the fuzzers.

```
cd tests
make -j zstreamtest fuzzer
./zstreamtest -T3m
./fuzzer -T3m
```

Those should give you a reasonable amount of confidence in your changes, along with the continuous integration passing when you put up your PR. After that we have oss-fuzz running continuously on our dev branch. So if something gets past all these tests, and code review, we should easily catch it before it merges into master.

Reply:
We should probably have better documentation for running tests for open source contributors. I'll file an issue.